pisa/utils/hyperplane.py

```diff
@@ -0,0 +1,628 @@
+
+        return on_axis_mask
+
+
+    def report(self,bin_idx=None) :
+        '''
+        String version of the hyperplane contents
+        '''
+
+        # Fit results
+        print(">>>>>> Fit parameters >>>>>>")
+        bin_indices = np.ndindex(self._binning.shape) if bin_idx is None else [bin_idx]
+        for bin_idx in bin_indices :
+            print("  Bin %s :" % (bin_idx,) )
+            print("     Intercept : %0.3g" % (self._intercept[bin_idx],) )
+            for sys_param in self._sys_params.values() :
+                print("     %s : %s" % ( sys_param.name, ", ".join([ "%0.3g"%sys_param._fit_params[(bin_idx,i,)] for i in range(sys_param._num_fit_params) ])) )
+        print("<<<<<< Fit parameters <<<<<<")
+
+
+
+class SysParam(object) :
+    '''
+    A class defining the systematic parameter in the hyperplane
+    Use constructs this by passing the functional form (as a function)
+    '''
+
+    def __init__(self,name,nominal_value,func_name,initial_fit_params=None) :
+
+        # Store basic members
+        self._name = name
+        self._nominal_value = nominal_value
+
+        # Handle functional form fit parameters
+        self._fit_params = None # Fir params container, not yet populated
+        self._initial_fit_params = initial_fit_params # The initial values for the fit parameters
+
+        # Record information relating to the fitting
+        self._fitted = False # Flag indicating whether fit has been performed
+        self._fitting_sys_dataset_values = None # The values of this sys param in each of the fitting datasets
+
+
+        #
+        # Init the functional form
+        #
+
+        # Get the function
+        self._func = self.get_func(func_name)
+
+        # Get the number of functional form parameters
+        # This is the functional form function parameters, excluding the systematic paramater and the output object
+        #TODO Does this support the GPU case?
+        self._num_fit_params = get_num_args(self._func) - 2
+
+        # Check and init the fit param initial values
+        #TODO Add support for per bin values?
+        if initial_fit_params is None :
+            # No values provided, use 0 for all
+            self._initial_fit_params = np.zeros(self._num_fit_params,dtype=FTYPE)
+        else :
+            # Use the provided initial values
+            self._initial_fit_params = np.array(self._initial_fit_params)
+            assert self._initial_fit_params.size == self._num_fit_params, "'initial_fit_params' should have %i values, found %i" % (self._num_fit_params,self._initial_fit_params.size)
+
+
+    def get_func(self,func_name) :
+        '''
+        Find the function defining the hyperplane functional form.
+
+        User specifies this by it's string name, which must correspond to one 
+        of the pre-defined functions.
+        '''
+
+        assert isinstance(func_name,basestring), "'func_name' must be a string"
```

Comment:
Something to get in the habit of doing is using `from six import string_types` and use `isinstance(func_name, string_types)` instead of `basestring` with the idea in mind that we're gonna have to be python3 compatible within a year or so. We need to make this change throughout all of PISA, but start thinking for new code in terms of making it py3 compatible.